################################################################################
SHORT DESCRIPTION: 
################################################################################
Add a XSM/Flask v4v use permission check

################################################################################
LONG DESCRIPTION: 
################################################################################
Add a XSM/Flask v4v use permission check to control the ability to invoke any of
the v4v hypercalls.  The existing v4v send permission check only controls the
ability to send via v4v, but still allows use of other v4v hypercalls, and a
subset of the v4v send code path is still reachable prior to the send check. The
use permission check can be used to remove the v4v hypercall interface entirely
from the attack surface of the hypervisor for domains that have no legitimate
reason to use v4v.

This change depends on a separate change to xsm-policy to allow v4v use by
authorized domains.

Also update the xsm_v4v_send hook to follow the upstream convention of including
a XSM_DEFAULT_ARG parameter and returning the result of xsm_default_action()
when XSM is disabled; this convention was introduced in xen commit
0d7f18b01f69c6b89aa3654bd2b11e24f41aaf71 but the v4v XSM send hook was
apparently not updated for it.  This has no impact in OpenXT since XSM is always
enabled, but makes it consistent with upstream.

################################################################################
CHANGELOG 
################################################################################
Author: Stephen Smalley <sds@tycho.nsa.gov>

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
--- a/xen/common/v4v.c
+++ b/xen/common/v4v.c
@@ -1747,7 +1747,7 @@ v4v_send (struct domain *src_d, v4v_addr
     }
 
   /* XSM: verify if src is allowed to send to dst */
-  if (xsm_v4v_send(src_d, dst_d) != 0)
+  if (xsm_v4v_send(XSM_HOOK, src_d, dst_d) != 0)
     {
       printk(KERN_ERR "V4V: XSM REJECTED %i -> %i\n",
              src_addr->domain, dst_addr->domain);
@@ -1858,7 +1858,7 @@ v4v_sendv (struct domain *src_d, v4v_add
     }
 
   /* XSM: verify if src is allowed to send to dst */
-  if (xsm_v4v_send(src_d, dst_d) != 0)
+  if (xsm_v4v_send(XSM_HOOK, src_d, dst_d) != 0)
     {
       printk(KERN_ERR "V4V: XSM REJECTED %i -> %i\n",
              src_addr->domain, dst_addr->domain);
@@ -1938,7 +1938,11 @@ do_v4v_op (int cmd, XEN_GUEST_HANDLE (vo
            XEN_GUEST_HANDLE (void) arg3, uint32_t arg4, uint32_t arg5)
 {
   struct domain *d = current->domain;
-  long rc = -EFAULT;
+  long rc;
+
+  rc = xsm_v4v_use(XSM_HOOK, d);
+  if (rc)
+      return rc;
 
 #ifdef V4V_DEBUG
 
@@ -1949,6 +1953,8 @@ do_v4v_op (int cmd, XEN_GUEST_HANDLE (vo
 #endif
 
   domain_lock (d);
+
+  rc = -EFAULT;
   switch (cmd)
     {
     case V4VOP_register_ring:
--- a/xen/include/xsm/dummy.h
+++ b/xen/include/xsm/dummy.h
@@ -724,9 +724,16 @@ static XSM_INLINE int xsm_dm_op(XSM_DEFA
     return xsm_default_action(action, current->domain, d);
 }
 
-static XSM_INLINE int xsm_v4v_send(struct domain *d, struct domain *t)
+static XSM_INLINE int xsm_v4v_send(XSM_DEFAULT_ARG struct domain *d, struct domain *t)
 {
-    return 0;
+    XSM_ASSERT_ACTION(XSM_HOOK);
+    return xsm_default_action(action, d, t);
+}
+
+static XSM_INLINE int xsm_v4v_use(XSM_DEFAULT_ARG struct domain *d)
+{
+    XSM_ASSERT_ACTION(XSM_HOOK);
+    return xsm_default_action(action, d, d);
 }
 
 #endif /* CONFIG_X86 */
--- a/xen/include/xsm/xsm.h
+++ b/xen/include/xsm/xsm.h
@@ -183,6 +183,7 @@ struct xsm_operations {
     int (*xen_version) (uint32_t cmd);
     int (*domain_resource_map) (struct domain *d);
     int (*v4v_send) (struct domain *dom1, struct domain *dom2);
+    int (*v4v_use) (struct domain *d);
 #ifdef CONFIG_ARGO
     int (*argo_enable) (const struct domain *d);
     int (*argo_register_single_source) (const struct domain *d,
@@ -693,11 +694,16 @@ static inline int xsm_dm_op(xsm_default_
     return xsm_ops->dm_op(d);
 }
 
-static inline int xsm_v4v_send(struct domain *d1, struct domain *d2)
+static inline int xsm_v4v_send(xsm_default_t def, struct domain *d1, struct domain *d2)
 {
     return xsm_ops->v4v_send(d1,d2);
 }
 
+static inline int xsm_v4v_use(xsm_default_t def, struct domain *d)
+{
+    return xsm_ops->v4v_use(d);
+}
+
 #endif /* CONFIG_X86 */
 
 static inline int xsm_xen_version (xsm_default_t def, uint32_t op)
--- a/xen/xsm/dummy.c
+++ b/xen/xsm/dummy.c
@@ -154,6 +154,7 @@ void __init xsm_fixup_ops (struct xsm_op
     set_to_dummy_if_null(ops, xen_version);
     set_to_dummy_if_null(ops, domain_resource_map);
     set_to_dummy_if_null(ops, v4v_send);
+    set_to_dummy_if_null(ops, v4v_use);
 #ifdef CONFIG_ARGO
     set_to_dummy_if_null(ops, argo_enable);
     set_to_dummy_if_null(ops, argo_register_single_source);
--- a/xen/xsm/flask/hooks.c
+++ b/xen/xsm/flask/hooks.c
@@ -1724,6 +1724,11 @@ static int flask_v4v_send(struct domain
     return domain_has_perm(dom1, dom2, SECCLASS_V4V, V4V__SEND);
 }
 
+static int flask_v4v_use(struct domain *d)
+{
+    return current_has_perm(d, SECCLASS_V4V, V4V__USE);
+}
+
 long do_flask_op(XEN_GUEST_HANDLE_PARAM(xsm_op_t) u_flask_op);
 int compat_flask_op(XEN_GUEST_HANDLE_PARAM(xsm_op_t) u_flask_op);
 
@@ -1859,6 +1864,7 @@ static struct xsm_operations flask_ops =
     .xen_version = flask_xen_version,
     .domain_resource_map = flask_domain_resource_map,
     .v4v_send = flask_v4v_send,
+    .v4v_use = flask_v4v_use,
 #ifdef CONFIG_ARGO
     .argo_enable = flask_argo_enable,
     .argo_register_single_source = flask_argo_register_single_source,
--- a/xen/xsm/flask/policy/access_vectors
+++ b/xen/xsm/flask/policy/access_vectors
@@ -535,4 +535,5 @@ class version
 class v4v
 {
     send
+    use
 }
